src/vmm/src/builder.rs

```diff
@@ -763,39 +767,657 @@ pub mod tests {
+                .unwrap();
+        let i8042_reset_evfd = EventFd::new(libc::EFD_NONBLOCK).unwrap();
+        PortIODeviceManager::new(serial, i8042_reset_evfd).unwrap()
+    }
+
+    fn default_kernel_cmdline() -> Cmdline {
+        let mut kernel_cmdline = kernel::cmdline::Cmdline::new(4096);
+        kernel_cmdline.insert_str(DEFAULT_KERNEL_CMDLINE).unwrap();
+        kernel_cmdline
+    }
+
+    fn default_boot_config() -> BootConfig {
+        BootConfig {
+            cmdline: default_kernel_cmdline(),
+            kernel_file: default_kernel_file(),
+        }
+    }
+
+    fn default_vmm(event_manager: &mut EventManager) -> Vmm {
+        let guest_memory = create_guest_memory(128).unwrap();
+        let kernel_cmdline = default_kernel_cmdline();
+
+        let exit_evt = EventFd::new(libc::EFD_NONBLOCK)
+            .map_err(Error::EventFd)
+            .map_err(StartMicrovmError::Internal)
+            .unwrap();
+
+        let vm = setup_kvm_vm(&guest_memory).unwrap();
+        let mmio_device_manager = default_mmio_device_manager();
+        #[cfg(target_arch = "x86_64")]
+        let pio_device_manager = default_portio_device_manager(event_manager);
+
+        Vmm {
+            events_observer: Some(Box::new(SerialStdin::get())),
+            guest_memory,
+            kernel_cmdline,
+            vcpus_handles: Vec::new(),
+            exit_evt,
+            vm,
+            mmio_device_manager,
+            #[cfg(target_arch = "x86_64")]
+            pio_device_manager,
+            seccomp_level: SECCOMP_LEVEL_ADVANCED,
+        }
+    }
+
+    fn create_x86_64_vcpus_with_template(
+        template: Option<CpuFeaturesTemplate>,
+        vcpu_count: u8,
+    ) -> crate::Result<Vec<Vcpu>> {
         let mut event_manager = EventManager::new().expect("Unable to create EventManager");
-        setup_serial_device(
-            &mut event_manager,
-            Box::new(read_handle),
-            Box::new(io::stdout()),
+        let guest_memory = create_guest_memory(128).unwrap();
+        let mut vm = setup_kvm_vm(&guest_memory).unwrap();
+        setup_interrupt_controller(&mut vm).unwrap();
+        let vcpu_config = VcpuConfig {
+            vcpu_count,
+            ht_enabled: false,
+            cpu_template: template,
+        };
+        let boot_config = default_boot_config();
+        let entry_addr = load_kernel(&boot_config, &guest_memory).unwrap();
+        let pio_dev_manager = default_portio_device_manager(&mut event_manager);
+        create_vcpus_x86_64(
+            &vm,
+            &vcpu_config,
+            &guest_memory,
+            entry_addr,
+            TimestampUs::default(),
+            &pio_dev_manager.io_bus,
+        )
+    }
+
+    #[test]
+    fn test_stdin_wrapper() {
+        let wrapper = SerialStdin(io::stdin());
+        assert_eq!(wrapper.as_raw_fd(), io::stdin().as_raw_fd())
+    }
+
+    #[test]
+    #[cfg(target_arch = "x86_64")]
+    fn test_create_vcpus_x86_with_t2_template() {
```

Comment:
I don't think we should test the different templates in this file. The tests here should just verify the correctness of the `create_vcpus_x86_64()` function not also verify the correctness of the functions it calls.

To that purpose, I believe it should be enough to just verify a valid `Vec<Vcpu>` of the right length is returned.